File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-5.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack = []
        max_area = 0
        heights.append(0)

        for i,h in enumerate(heights):
                while stack and h < heights[stack[-1]]:
                        height = heights[stack[-1]]
                        stack.pop()

                        if not stack:
                                width = i
                        else :
                                width = (i - stack[-1]) - 1
                        max_area = max(max_area , height*width)
                stack.append(i)
        return max_area
                        

File: Data Structures & Algorithms/max-water-container/submission-5.js
class Solution {
    /**
     * @param {number[]} heights
     * @return {number}
     */
    maxArea(heights) {
          const size = heights.length;
          let area = 0;
          let left =0;
          let right = size-1;

          while(left<right){
            const height = Math.min(heights[left],heights[right]);
            const width =  (right-left);
            area = Math.max(area, height*width);

            if (heights[left] < heights[right]) {
                left++;
            } else {
                right--;
            }
          }

        return area;
    }
}
